src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1374,41 +1396,80 @@ RouterJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
 		return job;
 	}
 
-	pg_get_query_def(originalQuery, queryString);
-
-	if (originalQuery->commandType == CMD_SELECT)
+	/*
+	 * If it's a multiple shard update/delete query we need to create task for
+	 * each shard interval.
+	 */
+	if (*multiShardModifyQuery)
 	{
-		task = CreateTask(ROUTER_TASK);
+		List *taskList = NIL;
+		ListCell *relationShardCell = NULL;
+		int taskId = 1;
+
+		foreach(relationShardCell, relationShardList)
+		{
+			RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+			List *relationShardList = list_make1(relationShard);
+			Query *copiedQuery = copyObject(originalQuery);
+			Task *task = CreateTask(MODIFY_TASK);
+			StringInfo shardQueryString = makeStringInfo();
+
+			if (!requiresMasterEvaluation)
+			{
+				UpdateRelationToShardNames((Node *) copiedQuery, relationShardList);
+				pg_get_query_def(copiedQuery, shardQueryString);
+			}
+
+			task->taskId = taskId++;
+			task->anchorShardId = relationShard->shardId;
+			task->queryString = shardQueryString->data;
+			task->relationShardList = relationShardList;
+
+			taskList = lappend(taskList, task);
+		}
+
+
+		job->taskList = taskList;
```

Comment:
This function is still not very readable.

Maybe something like:
```
if (originalQuery->commandType == CMD_SELECT)
{
    taskList = SingleShardSelectTaskList(originalQuery, relationShardList, placementList);
}
else if (*multiShardModifyQuery) // <-- can be a regular variable after removing the parameter
{
    taskList = MultiShardModifyTaskList(originalQuery, relationShardList);
}
else
{
    taskList = SingleShardModifyTaskList(originalQuery, relationShardList);
}
```

I'm also wondering whether we can merge the latter two.